server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/S3DumbBlobStore.java

```diff
@@ -0,0 +1,264 @@
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.core.BytesWrapper;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.async.AsyncResponseTransformer;
+import software.amazon.awssdk.core.async.SdkPublisher;
+import software.amazon.awssdk.core.exception.SdkClientException;
+import software.amazon.awssdk.http.nio.netty.NettyNioAsyncHttpClient;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.model.BucketAlreadyOwnedByYouException;
+import software.amazon.awssdk.services.s3.model.DeleteObjectsResponse;
+import software.amazon.awssdk.services.s3.model.GetObjectResponse;
+import software.amazon.awssdk.services.s3.model.ListBucketsResponse;
+import software.amazon.awssdk.services.s3.model.ListObjectsResponse;
+import software.amazon.awssdk.services.s3.model.NoSuchBucketException;
+import software.amazon.awssdk.services.s3.model.NoSuchKeyException;
+import software.amazon.awssdk.services.s3.model.ObjectIdentifier;
+import software.amazon.awssdk.services.s3.model.PutObjectRequest;
+import software.amazon.awssdk.services.s3.model.S3Object;
+
+public class S3DumbBlobStore implements DumbBlobStore {
+
+    private static final int CHUNK_SIZE = 1024 * 1024;
+    private static final int EMPTY_BUCKET_BATCH_SIZE = 1000;
+    private static final int FILE_THRESHOLD = 1024 * 100;
+    private static final Duration FIRST_BACK_OFF = Duration.ofMillis(100);
+    private static final Duration FOREVER = Duration.ofMillis(Long.MAX_VALUE);
+    private static final boolean LAZY = false;
+    private static final int MAX_RETRIES = 5;
+
+    private final S3AsyncClient client;
+
+    @Inject
+    S3DumbBlobStore(AwsS3AuthConfiguration configuration, Region region) {
+        client = S3AsyncClient.builder()
+            .credentialsProvider(StaticCredentialsProvider.create(
+                AwsBasicCredentials.create(configuration.getAccessKeyId(), configuration.getSecretKey())))
+            .httpClient(NettyNioAsyncHttpClient.builder()
+                .maxConcurrency(100)
+                .maxPendingConnectionAcquires(10_000)
+                .build())
+            .endpointOverride(URI.create(configuration.getEndpoint()))
+            .region(region.asAws())
+            .build();
+    }
+
+    public void close() {
+        client.close();
+    }
+
+    @Override
+    public InputStream read(BucketName bucketName, BlobId blobId) throws ObjectStoreIOException, ObjectNotFoundException {
+        return getObject(bucketName, blobId)
+            .map(response -> ReactorUtils.toInputStream(response.flux))
+            .onErrorMap(NoSuchBucketException.class, e -> new ObjectNotFoundException("Bucket not found " + bucketName, e))
+            .onErrorMap(NoSuchKeyException.class, e -> new ObjectNotFoundException("Blob not found " + bucketName, e))
+            .block();
+    }
+
+    private static class FluxResponse {
```

Comment:
class ordering

Reply:
As it's an implementation detail, I decided to put it close to its usage.